docs/toctoc/guides/token-based-authentication-flow.md

```diff
@@ -0,0 +1,210 @@
+---
+id: token-based-authentication-flow
+title: Token-based Authentication Flow
+---
+
+This flow allows to implement the common signup, login and authentication
+functionalies of a web application.
+
+## Authentication domains
+
+The flow defines and orchestrates two
+[authentication domains](toctoc/concepts/functional-model.md):
+
+- **`LoginDomain`**: an authentication domain whose credential is the `Login`
+  data type.
+- **`AccessTokenDomain`**: an authentication domain whose credential is the
+  `AccessToken` data type.
+
+The `Login` data type is defined by a username and a password, while the
+`AccessToken` data type is defined by a value -- which is usually a randomly
+generated string -- and an expiration date.
+
+## Functionalities
+
+The flow is implemented by the `TokenBasedAuthenticationFGlow` class, which
+exposes these functionalities:
+
+- `registerSubjectLogin`: registers the `Login` of a `Subject` in the
+  `LoginDomain`. This is typically used when signing up a new user.
+
+- `exchangeForTokens`: exchanges a `Login` for an `AccessToken`. This is
+  typically used when a user logs into the system providing username and
+  password, and they receive an `AccessToken` that they will use for
+  authenticating further requests.
+
+- `validateToken`: checks whether an `AccessToken` is valid and returns the
+  authenticated `Subject` if true.
+
+- `unregisterToken`: invalidates an `AccessToken`. This is typically used when a
+  user logs out from a device.
+
+- `unregisterAllSubjectTokens`: invalidates all `AccessToken` of a `Subject`.
+  This can be used to implement security features, such as logging out from all
+  active sessions in case the password is suspected to be compromised or upon a
+  change password request.
+
+- `unregisterLogin`: invalidates a `Login`. This can be used in case of a change
+  password request in an app that wants to guarantee a single login at the time
+  per user.
+
+- `unregisterAllSubjectLogins`: invalidates all `Login` of a `Subject`. This is
```

Comment:
@gabro the `update` of the Login of a Subject is not implemented yet, right?

https://github.com/buildo/retro/issues/12

Reply:
Correct, not implemented yet.